src/encode.rs

```diff
@@ -0,0 +1,188 @@
+    let mut encoder = Cursor::new(vec![]);
+    data.consensus_encode(&mut encoder).unwrap();
+    encoder.into_inner()
+}
+
+/// Encode an object into a hex-encoded string
+pub fn serialize_hex<T: Encodable + ?Sized>(data: &T) -> String {
+    ::bitcoin::hashes::hex::ToHex::to_hex(&serialize(data)[..])
+}
+
+/// Deserialize an object from a vector, will error if said deserialization
+/// doesn't consume the entire vector.
+pub fn deserialize<'a, T: Decodable>(data: &'a [u8]) -> Result<T, Error> {
+    let (rv, consumed) = deserialize_partial(data)?;
+
+    // Fail if data are not consumed entirely.
+    if consumed == data.len() {
+        Ok(rv)
+    } else {
+        Err(Error::ParseFailed("data not consumed entirely when explicitly deserializing"))
+    }
+}
+
+/// Deserialize an object from a vector, but will not report an error if said deserialization
+/// doesn't consume the entire vector.
+pub fn deserialize_partial<'a, T: Decodable>(data: &'a [u8]) -> Result<(T, usize), Error> {
+    let mut decoder = Cursor::new(data);
+    let rv = Decodable::consensus_decode(&mut decoder)?;
+    let consumed = decoder.position() as usize;
+
+    Ok((rv, consumed))
+}
+
+/// Implement Elements encodable traits for Bitcoin encodable types.
+macro_rules! impl_upstream {
+    ($type: ty) => {
+        impl Encodable for $type {
+            fn consensus_encode<W: io::Write>(&self, mut e: W) -> Result<usize, Error> {
+                Ok(btcenc::Encodable::consensus_encode(self, &mut e)?)
+            }
+        }
+
+        impl Decodable for $type {
+            fn consensus_decode<D: io::Read>(mut d: D) -> Result<Self, Error> {
+                Ok(btcenc::Decodable::consensus_decode(&mut d)?)
+            }
+        }
+    };
+}
+impl_upstream!(u8);
+impl_upstream!(u32);
+impl_upstream!(u64);
+impl_upstream!([u8; 32]);
+impl_upstream!(Vec<u8>);
+impl_upstream!(Vec<Vec<u8>>);
+impl_upstream!(btcenc::VarInt);
+impl_upstream!(::bitcoin::blockdata::script::Script);
+impl_upstream!(::bitcoin::hashes::sha256d::Hash);
+
+// Vectors
+macro_rules! impl_vec {
+    ($type: ty) => {
+        impl Encodable for Vec<$type> {
+            #[inline]
+            fn consensus_encode<S: io::Write>(&self, mut s: S) -> Result<usize, Error> {
+                let mut len = 0;
+                len += btcenc::VarInt(self.len() as u64).consensus_encode(&mut s)?;
+                for c in self.iter() {
+                    len += c.consensus_encode(&mut s)?;
+                }
+                Ok(len)
+            }
+        }
+
+        impl Decodable for Vec<$type> {
+            #[inline]
+            fn consensus_decode<D: io::Read>(mut d: D) -> Result<Self, Error> {
+                let len = btcenc::VarInt::consensus_decode(&mut d)?.0;
+                let byte_size = (len as usize)
+                    .checked_mul(mem::size_of::<$type>())
+                    .ok_or(self::Error::ParseFailed("Invalid length"))?;
+                if byte_size > btcenc::MAX_VEC_SIZE {
```

Comment:
and we should `pub use bitcoin::consensus::encode::MAX_VEC_SIZE` in this module

Reply:
Ah yeah instead of redefine. Agree with that.